# peakScout <a href="https://github.com/vandydata/peakScout"><img src="assets/logo.svg" align="right" height="150"  alt="peakScout website"/></a>

<!-- badges: start -->
[![peakScout](https://github.com/vandydata/peakScout/actions/workflows/python.yaml/badge.svg)](https://github.com/vandydata/peakScout/actions/workflows/python.yaml)
[![License: AGPLv3](https://img.shields.io/badge/License-AGPL%20v3-blue.svg)](https://www.gnu.org/licenses/agpl-3.0)
<!-- badges: end -->

peakScout is a user-friendly and reversible peak-to-gene translator for genomic peak calling results

> Please cite CITATION if you use peakScout, thank you!

## Overview

PeakScout is a bioinformatics tool designed to bridge the gap between genomic peak data and gene annotations, enabling researchers to understand the relationship between regulatory elements and their target genes. At its core, peakScout processes genomic peak files generated by popular peak callers like MACS2 and SEACR and maps them to nearby genes using reference genome annotations. 

peakScount performs:
- **Peak-to-Gene Mapping**: This function identifies the nearest genes to each peak, allowing researchers to infer which genes might be regulated by specific genomic regions. Users can specify how many nearest genes (k) they want to retrieve for each peak.
- **Gene-to-Peak Mapping**: Conversely, this function finds the nearest peaks to a list of genes, helping researchers identify potential regulatory elements that may influence gene expression.

peakScout expects two inputs: 
1. A **peak file** (in BED6 format or as output from MACS2 or SEACR) 
2. A **reference GTF file** containing gene annotations. The tool can decompose the GTF file into chromosome-specific collections of genomic features, which are then used to perform bidirectional mapping between peaks and genes.

peakScount can be run:
- **Command line**: peakScout is designed to be run from the command line, making it accessible for users comfortable with terminal operations.
- **Cloud computing**: for instanct access web access, we have set up peakScout in the cloud - https://vandydata.github.io/peakScout.



## Installation

## From source

These instructions should generally work without modification in linux-based environments. If you are using Windows, we strongly recommend you use WSL2 to have a Linux environment within Windows.
```bash
# 1. Clone the Repository
git clone https://github.com/vandydata/peakScout.git

# 2. Make the Script Executable
cd peakScout; chmod +x src/peakScout

# 3. Add to Path
# Alternatively, edit your `~/.bashrc` to make this change permanent, but be sure 
# to include the complete path in the file itself, not the `$(pwd)`. 
export PATH="$PATH:$(pwd)/src"

# 4. Set up virtual environment and install dependencies
# with venv
python3 -m venv peakscout
source peakscout/bin/activate
pip3 install -r requirements.txt

# OR with uv
uv venv peakscout
source peakscout/bin/activate
uv pip install -r requirements.txt
```

### Docker & singularity containers

We have made available a Docker image for peakScout, which can be run as follows:

```bash
docker run -it --rm jpcartailler/peakscout:latest peakScout --help
```

For singularity, you can convert the Docker image to a Singularity image and run it as follows:

```bash
singularity pull docker://jpcartailler/peakscout:latest
singularity exec peakscout_latest.sif peakScout --help
```



## Usage

### Decomposing Reference GTF
To decompose a reference GTF file so that it can be used by peakScout, run the following command
```bash
peakScout decompose --ref_dir /path/to/where/outputs/stored --species species of gtf --gtf_ref /path/to/gtf/file
```

Specific example:

```bash
peakScout decompose \
--ref_dir reference/ \
--species mm39 \
--gtf_ref reference/gencode.vM37.basic.annotation.gtf
```

A directory called `reference/mm39` will be created and you will use the `mm39` as the species name in other peakScout operations.

### Finding Nearest Genes

Once a reference GTF has been decomposed, you can use the decomposition to find the nearest genes to your peaks. Peak files can be MACS2, SEACR outputs, or standard BED6 format files and can be Excel sheets or BED files.

Run the following command to create an Excel sheet containing the nearest k genes to your peaks
```bash
peakScout peak2gene --peak_file /path/to/peak/file --peak_type MACS2/SEACR/BED6 --species species of gtf --k number of nearest genes --ref_dir /path/to/reference/directory --output_name name of output file --o /path/to/save/output --output_type csv/xslx
```

Specific example:

```bash
peakScout peak2gene \
--peak_file test/test_MACS2.bed \
--peak_type MACS2 \
--species mm39 \
--k 2 \
--ref_dir reference/mm39 \
--output_name peakScout_test_MACS2 \
--o my_output_dir \
--output_type xslx
```

Specific example:

```bash
peakScout peak2gene \
--peak_file test/test_MACS2.bed \
--peak_type MACS2 \
--species mm39 \
--k 2 \
--ref_dir reference/mm39 \
--output_name peakScout_test_MACS2 \
--o my_output_dir \
--output_type xslx
```

### Finding Nearest Peaks

Once a reference GTF has been decomposed, you can use the decomposition to find the nearest peaks to a set of genes. Peak files can be MACS2, SEACR outputs, or standard BED6 format files and can be Excel sheets or BED files. Gene names should be in a single column CSV file with no header.

Run the following command to create an Excel sheet containing the nearest k peaks to your genes
```bash
peakScout gene2peak --peak_file /path/to/peak/file --peak_type MACS2/SEACR/BED6 --gene_file /path/to/gene/file --species species of gtf --k number of nearest peaks --ref_dir /path/to/reference/directory --output_name name of output file --o /path/to/save/output --output_type csv/xslx
```

Specific example:
```bash
peakScout gene2peak --peak_file /path/to/peak/file --peak_type MACS2/SEACR/BED6 --gene_file /path/to/gene/file --species species of gtf --k number of nearest peaks --ref_dir /path/to/reference/directory --output_name name of output file --o /path/to/save/output --output_type csv/xslx
```

## peakScout ready-made references for common organisms

For your convenience, we have prepared reference files for common organisms, generated by `src/utils/decompose-common-organisms.sh`. Source files are the GTFs below and downloadable peakScout reference files are the S3 links.

| Species | GTF | S3 |
|---------|-----|-----|
| arabidopsis_TAIR10 | [GTF](https://ftp.ebi.ac.uk/ensemblgenomes/pub/plants/current/gtf/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.61.gtf.gz) | [S3](https://cds-peakscout-public.s3.us-east-1.amazonaws.com/arabidopsis_TAIR10.tar.zst) |
| fly_BDGP6.54 | [GTF](https://ftp.ensembl.org/pub/current_gtf/drosophila_melanogaster/Drosophila_melanogaster.BDGP6.32.114.gtf.gz) | [S3](https://cds-peakscout-public.s3.us-east-1.amazonaws.com/fly_BDGP6.54.tar.zst) |
| frog_v10.1 | [GTF](https://ftp.ensembl.org/pub/current_gtf/xenopus_tropicalis/Xenopus_tropicalis.UCB_Xtro_10.0.114.gtf.gz) | [S3](https://cds-peakscout-public.s3.us-east-1.amazonaws.com/frog_v10.1.tar.zst) |
| human_hg19 | [GTF](https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_10/gencode.v10.annotation.gtf.gz) | [S3](https://cds-peakscout-public.s3.us-east-1.amazonaws.com/human_hg19.tar.zst) |
| human_hg38 | [GTF](https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_48/gencode.v48.basic.annotation.gtf.gz) | [S3](https://cds-peakscout-public.s3.us-east-1.amazonaws.com/human_hg38.tar.zst) |
| mouse_mm10 | [GTF](https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gtf.gz) | [S3](https://cds-peakscout-public.s3.us-east-1.amazonaws.com/mouse_mm10.tar.zst) |
| mouse_mm39 | [GTF](https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M37/gencode.vM37.basic.annotation.gtf.gz) | [S3](https://cds-peakscout-public.s3.us-east-1.amazonaws.com/mouse_mm39.tar.zst) |
| pig_Sscrofa11.1 | [GTF](https://ftp.ensembl.org/pub/release-114/gtf/sus_scrofa/Sus_scrofa.Sscrofa11.1.114.gtf.gz) | [S3](https://cds-peakscout-public.s3.us-east-1.amazonaws.com/pig_Sscrofa11.1.tar.zst) |
| worm_WBcel235 | [GTF](https://ftp.ebi.ac.uk/ensemblgenomes/pub/metazoa/current/gtf/caenorhabditis_elegans/Caenorhabditis_elegans.WBcel235.61.gtf.gz) | [S3](https://cds-peakscout-public.s3.us-east-1.amazonaws.com/worm_WBcel235.tar.zst) |
| yeast_R64-1-1 | [GTF](https://ftp.ebi.ac.uk/ensemblgenomes/pub/fungi/current/gtf/saccharomyces_cerevisiae/Saccharomyces_cerevisiae.R64-1-1.61.gtf.gz) | [S3](https://cds-peakscout-public.s3.us-east-1.amazonaws.com/yeast_R64-1-1.tar.zst) |
| zebrafish_GRCz11 | [GTF](https://ftp.ensembl.org/pub/current_gtf/danio_rerio/Danio_rerio.GRCz11.114.gtf.gz) | [S3](https://cds-peakscout-public.s3.us-east-1.amazonaws.com/zebrafish_GRCz11.tar.zst) |


## FAQ

### What is WSL2 and how do I install it?

Start at https://learn.microsoft.com/en-us/windows/wsl/install and then Google for your issues. We cannot provide support for this. 

### How to get a GTF file?

Ensembl and Gencode have them. For example, go to https://www.gencodegenes.org/mouse/release_M37.html and select the GTF you'd like to decompose, then:
```
mkdir reference
cd reference
wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M37/gencode.vM37.basic.annotation.gtf.gz
gunzip gencode.vM37.basic.annotation.gtf.gz
```


## Notes

Genomes to target: mm10, mm39, hg19, hg38

Get annotations from Gencode, i.e. M25 for mm10 (which is aka GRCm38)

From gencode files, we can get chr, start, end, feature name, feature type (`gene`, `transcript`, etc), and biotype (eg. `gene_type "TEC";`)

Write a parser for the GTF data and at least get it into the lowest common denominator currently need, optimizations can come later

