# peakScout

## Installation

### 1. Clone the Repository

```bash
git clone https://github.com/vandydata/peakScout.git
cd peakScout/src
```

### 2. Make the Script Executable
```bash
chmod +x peakScout
```

### 3. Add to Path
```bash
export PATH="$PATH:$(pwd)"
```

Alternatively, add this line to ~/.bashrc to make this change permanent

### 4. Set Up Virtual Environment
```bash
python3 -m venv peakscout
source peakscout/bin/activate
pip3 install -r requirements.txt
```

## Usage

### Decomposing Reference GTF
To decompose a reference GTF file so that it can be used by peakScout, run the following command
```bash
peakScout decompose --ref_dir /path/to/where/outputs/stored --species species of gtf --gtf_ref /path/to/gtf/file
```

### Finding Nearest Genes
Once a reference GTF has been decomposed, you can use the decomposition to find the nearest genes to your peaks. Peak files can be MACS2 or SEACR outputs and can be Excel sheets or BED files.

Run the following command to create an Excel sheet containing the nearest k genes to your peaks
```bash
peakScout peak2gene --peak_file /path/to/peak/file --peak_type MACS2/SEACR --species species of gtf --k number of nearest genes --ref_dir /path/to/reference/directory --output_name name of output file --o /path/to/save/output --output_type csv/xslx
```

### Finding Nearest Peaks
Once a reference GTF has been decomposed, you can use the decomposition to find the nearest peaks to a set of genes. Peak files can be MACS2 or SEACR outputs and can be Excel sheets or BED files. Gene names should be in a single column CSV file with no header.

Run the following command to create an Excel sheet containing the nearest k peaks to your genes
```bash
peakScout gene2peak --peak_file /path/to/peak/file --peak_type MACS2/SEACR --gene_file /path/to/gene/file --species species of gtf --k number of nearest peaks --ref_dir /path/to/reference/directory --output_name name of output file --o /path/to/save/output --output_type csv/xslx
```

## Decomposed references for common organisms

For your convenience, we have prepared decomposed GTF files for common organisms, generated by `src/utils/decompose-common-organisms.sh` from the following GTF files:

* `arabidopsis_TAIR10` - https://ftp.ebi.ac.uk/ensemblgenomes/pub/plants/current/gtf/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.61.gtf.gz
* `fly_BDGP6.54` - https://ftp.ensembl.org/pub/current_gtf/drosophila_melanogaster/Drosophila_melanogaster.BDGP6.32.114.gtf.gz
* `frog_v10.1` - https://ftp.ensembl.org/pub/current_gtf/xenopus_tropicalis/Xenopus_tropicalis.UCB_Xtro_10.0.114.gtf.gz
* `human_hg38` - https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_48/gencode.v48.basic.annotation.gtf.gz
* `human_hg19` - https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_10/gencode.v10.annotation.gtf.gz
* `mouse_mm39` - https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M37/gencode.vM37.basic.annotation.gtf.gz
* `mouse_mm10` - https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gtf.gz
* `pig_Sscrofa11.1` - https://ftp.ensembl.org/pub/release-114/gtf/sus_scrofa/Sus_scrofa.Sscrofa11.1.114.gtf.gz
* `worm_WBcel235` - https://ftp.ebi.ac.uk/ensemblgenomes/pub/metazoa/current/gtf/caenorhabditis_elegans/Caenorhabditis_elegans.WBcel235.61.gtf.gz
* `yeast_R64-1-1` - https://ftp.ebi.ac.uk/ensemblgenomes/pub/fungi/current/gtf/saccharomyces_cerevisiae/Saccharomyces_cerevisiae.R64-1-1.61.gtf.gz
* `zebrafish_GRCz11` - https://ftp.ensembl.org/pub/current_gtf/danio_rerio/Danio_rerio.GRCz11.114.gtf.gz

You can download these from a publix AWS S3 bucket as ZSTD-compressed files, which can be decompressed with `zstd -d`:

* `s3://peakscout/common-organisms/arabidopsis_TAIR10.zst`
* `s3://peakscout/common-organisms/worm_WBcel235.zst`
* etc...



## Notes

Genomes to target: mm10, mm39, hg19, hg38

Get annotations from Gencode, i.e. M25 for mm10 (which is aka GRCm38)

From gencode files, we can get chr, start, end, feature name, feature type (`gene`, `transcript`, etc), and biotype (eg. `gene_type "TEC";`)

Write a parser for the GTF data and at least get it into the lowest common denominator currently need, optimizations can come later

